File: DB/MemoryLibrary.py
import mysql.connector as connector
from mysql.connector import errorcode
import csv

#  CREATE TABLE memorys (
#     id INT AUTO_INCREMENT PRIMARY KEY,                                            #   记忆的唯一标识符，自动递增；
#     memory VARCHAR(255) NOT NULL,                                                 #   记忆本身，必填字段；
#     definition TEXT,                                                              #   记忆的定义，可选字段；
#     context TEXT,                                                                 #   记忆出现的上下文，可选字段；
#     source VARCHAR(255),                                                          #   记忆来源，可选字段；
#     category VARCHAR(255),                                                        #   记忆所属的分类，可选字段；
#     created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,                               #   记忆的创建时间，自动生成；
#     updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP    #   记忆的更新时间，自动更新。
# );

# 数据库连接配置
config = {
    'user': 'root',
    'password': 'password',
    'host': 'localhost',
    'database': 'memory_db'
}

# 连接数据库
try:
    cnx = connector.connect(**config)
    print("Connected to database!")
except connector.Error as err:
    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
        print("Something is wrong with your username or password")
    elif err.errno == errorcode.ER_BAD_DB_ERROR:
        print("Database does not exist")
    else:
        print(err)

# 获取游标
cursor = cnx.cursor()


# 增加一条记忆
def add_memory(memory, definition='', context='', source='', category=''):
    add_query = ("INSERT INTO memorys "
                 "(memory, definition, context, source, category) "
                 "VALUES (%s, %s, %s, %s, %s)")
    data = (memory, definition, context, source, category)
    cursor.execute(add_query, data)
    cnx.commit()


# 根据记忆查询
def search_by_memory(memory):
    search_query = ("SELECT * FROM memorys "
                    "WHERE memory = %s")
    data = (memory,)
    cursor.execute(search_query, data)
    result = cursor.fetchall()
    return result


# 根据分类查询
def search_by_category(category):
    search_query = ("SELECT * FROM memorys "
                    "WHERE category = %s")
    data = (category,)
    cursor.execute(search_query, data)
    result = cursor.fetchall()
    return result


# 更新一条记忆
def update_memory(memory_id, memory, definition='', context='', source='', category=''):
    update_query = ("UPDATE memorys "
                    "SET memory=%s, definition=%s, context=%s, source=%s, category=%s "
                    "WHERE id=%s")
    data = (memory, definition, context, source, category, memory_id)
    cursor.execute(update_query, data)
    cnx.commit()


# 删除一条记忆
def delete_memory(memory_id):
    delete_query = ("DELETE FROM memorys "
                    "WHERE id=%s")
    data = (memory_id,)
    cursor.execute(delete_query, data)
    cnx.commit()


# 导出备份
def export_backup_memory(backup_file):
    with open(backup_file, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.writer(csvfile)
    writer.writerow(['id', 'memory', 'definition', 'context', 'source', 'category'])
    search_query = "SELECT * FROM memorys"
    cursor.execute(search_query)
    results = cursor.fetchall()
    for result in results:
        writer.writerow(result)
    print(f"Backup exported to {backup_file}")

def get_id_by_memory(memory):
    search_query = ("SELECT id FROM memorys "
                    "WHERE memory = %s")
    data = (memory,)
    cursor.execute(search_query, data)
    result = cursor.fetchone()
    if result:
        return result[0]
    else:
        return None

File: Codes/BackEnd/DataBase/MemoryLibrary.java
import java.util.ArrayList;
import java.util.List;

@Service
public class MemoryLibrary
{
    private final MemoryDAO memoryDAO;

    public MemoryLibrary(MemoryDAO memoryDAO)
    {
        this.memoryDAO = memoryDAO;
    }

    public boolean addMemory(String memory, String definition, String context, String source, String category)
    {
        // 调用 MemoryDAO 的 add_memory 方法
        return memoryDAO.addMemory(memory, definition, context, source, category);
    }

    public List<Memory> searchByMemory(String memory)
    {
        // 调用 MemoryDAO 的 search_by_memory 方法
        List<String[]> results = memoryDAO.searchByMemory(memory);

        // 将结果转换为 Memory 对象列表
        List<Memory> memories = new ArrayList<>();
        for (Object[] result : results)
        {
            Memory memoryObj = new Memory();
            memoryObj.setId((Long) result[0]);
            memoryObj.setMemory((String) result[1]);
            memoryObj.setDefinition((String) result[2]);
            memoryObj.setContext((String) result[3]);
            memoryObj.setSource((String) result[4]);
            memoryObj.setCategory((String) result[5]);
            memories.add(memoryObj);
        }

        return memories;
    }

    public List<Memory> searchByCategory(String category)
    {
        // 调用 MemoryDAO 的 search_by_category 方法
        List<String[]> results = memoryDAO.searchByCategory(category);

        // 将结果转换为 Memory 对象列表
        List<Memory> memories = new ArrayList<>();
        for (Object[] result : results)
        {
            Memory memoryObj = new Memory();
            memoryObj.setId((Long) result[0]);
            memoryObj.setMemory((String) result[1]);
            memoryObj.setDefinition((String) result[2]);
            memoryObj.setContext((String) result[3]);
            memoryObj.setSource((String) result[4]);
            memoryObj.setCategory((String) result[5]);
            memories.add(memoryObj);
        }

        return memories;
    }

    public boolean updateMemory(int memoryId, String memory, String definition, String context, String source,
            String category)
    {
        // 调用 MemoryDAO 的 update_memory 方法
        return memoryDAO.updateMemory(memoryId, memory, definition, context, source, category);
    }

    public boolean deleteMemory(int memoryId)
    {
        // 调用 MemoryDAO 的 delete_memory 方法
        return memoryDAO.deleteMemory(memoryId);
    }

    public boolean exportBackupMemory(String backupFile)
    {
        // 调用 MemoryDAO 的 export_backup_memory 方法
        return memoryDAO.exportBackupMemory(backupFile);
    }

    public int getIdByMemory(String memory)
    {
        // 调用 MemoryDAO 的 get_id_by_memory 方法
        Object result = memoryDAO.getIdByMemory(memory);
        if (result != null)
        {
            return (int) result;
        }
        else
        {
            return -1;
        }
    }
}
